### PR TITLE
Fix #2510: Yaml containing aliases rejected due to FasterXML bug

### LINKED ISSUE #2510: Yaml containing aliases rejected due to FasterXML bug
There is a long-standing bug in the way that FasterXML handles yaml containing aliases. https://github.com/FasterXML/jackson-dataformats-text/issues/98

As a consequence, anyone using perfectly valid yaml containing aliases (which may have been automatically generated via SnakeYaml) will get a mysterious error message like this:
```
15:26:16  hudson.remoting.ProxyException: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('id001')
15:26:16   at [Source: (BufferedInputStream); line: 49, column: 17] (through reference chain: io.fabric8.kubernetes.api.model.Pod["spec"]->io.fabric8.kubernetes.api.model.PodSpec["containers"]->java.util.ArrayList[1]->io.fabric8.kubernetes.api.model.Container["resources"]->io.fabric8.kubernetes.api.model.ResourceRequirements["requests"])
15:26:16  	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
15:26:16  	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1455)
15:26:16  	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1081)
15:26:16  	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)
15:26:16  	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromString(ValueInstantiator.java:258)
15:26:16  	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:357)
15:26:16  	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29)
15:26:16  	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
15:26:16  	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
15:26:16  	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:291)
15:26:16  	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:250)
15:26:16  	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
15:26:16  	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
15:26:16  	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
15:26:16  	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
15:26:16  	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4524)
15:26:16  	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3511)
15:26:16  	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:246)
```

Is the use of FasterXML strictly required? Would it be possible to simply use SnakeYaml directly for parsing, which seems to handle aliases properly?

### PR DESCRIPTION
**Pull request by rohanKanojia**
Fix #2510 
    Modify the way we deserialize Yaml objects. Load them into raw HashMap
    using SnakeYAML and then use Jackson to convert them to JSON and finally
    the required object

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [X] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java [hunk 1]
```diff
@@ -46,6 +48,7 @@ private Serialization() { }
     JSON_MAPPER.registerModule(new JavaTimeModule());
   }
   private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
+  private static final Yaml SNAKE_YAML_MAPPER = new Yaml();
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:51]: This class (`Yaml`) is not thread safe, should be instantiated for each request.

nit: it's not a mapper either.

The field should be removed. See proposed changes in lines 62-64 for the thread-safety issue.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java [hunk 2]
```diff
@@ -56,6 +59,10 @@ public static ObjectMapper yamlMapper() {
     return YAML_MAPPER;
   }
 
+  public static Yaml snakeYamlMapper() {
+    return SNAKE_YAML_MAPPER;
+  }
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:64]: https://github.com/fabric8io/kubernetes-client/pull/2646/files#r540041655
```suggestion
  public static Yaml snakeYaml() {
    return new Yaml();
  }
```

**Comment by manusa** [comment]
Since we want a consistent, deterministic behavior maybe instead of a fallback approach we could try something like:
**Deserialization**
- Load YAML into an **untyped** object using SnakeYaml.
- Serialize YAML into JSON
- Use Jackson to deserialize the JSON into the final type

This way we will **always** take advantage of custom deserializers specified with Jackson Annotations.

**Serialization**
No bug was reported for this process. I would really don't touch this.
- Object is serialized into YAML using Jackson

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java [hunk 3]
```diff
@@ -294,4 +294,21 @@ private static String readSpecFileFromInputStream(InputStream inputStream) {
       throw new RuntimeException("Unable to read InputStream." + e);
     }
   }
+
+  private static <T> T unmarshalYaml(InputStream is, TypeReference<T> type) throws JsonProcessingException {
+    final Yaml yaml = new Yaml();
+    // Load YAML into HashMap using SnakeYAML
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java:300]: nit: these comments are very redundant

**Review by manusa — APPROVED** [review]

**Comment by manusa** [comment]
[merge]

**Pull request merged**